Enable wallet to validate upon reorg

hansieodendaal opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
- This PR added the ability for a wallet to redo TXO and transaction validation protocols based on reorg information received from the base node when it receives new tip info data. Whenever a reorg is detected during the addition of blocks to the database backend, reorg information is updated in the blockchain database. The tip info struct has been updated to include the reorg information, and this is then provided to the wallet upon its chain metadata update cycle. In addition, two logical bugs were found and corrected in (a) the transaction validation protocol that should also validate broadcast transactions and (b) the broadcast monitoring protocol that should stop when transactions are not valid anymore.
- As an aside, fixed RPC connections that were being kept open when they should have been closed in the coinbase monitoring and transaction validation protocols.

_**Note:** The `ReorgInfo` struct contains two fields that are not being used at the moment; this is intentional and left for a follow-up optimization PR where the wallet can decide based on its own contents and chain height if TXO and transaction validation must be redone or not. The characteristics of hybrid mining with two or more algorithms are such that blockchain reorgs are numerous, especially 1 or 2 block deep reorgs. The current re-validation implementation will be triggered for all those reorgs._

## Motivation and Context
Wallets are not aware of transactions potentially being invalidated or validated if a base node they are connected to undergoes a reorg.

## How Has This Been Tested?
- Expanded cucumber test `Wallets monitoring coinbase and normal transactions after a reorg` to also verify transaction validity after a reorg.
- Verified in trace logs that the broadcast monitoring protocol stops after transactions become invalid.
- All unit test passed.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [X] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [X] I'm merging against the `development` branch.
* [X] I ran `cargo-fmt --all` before pushing.
* [X] I ran `cargo test` successfully before submitting my PR.
* [X] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

Commit:
Enable wallet to validate upon reorg

- Added reorg metadata to the chain metadata struct
  in the blockchain db
- Provide the reorg data to the wallet upon chain
  metadata update
- Wallet redo TXO and transaction validation protocols
  based on reorg information
- Fixed RPC connections being kept open; closing RPC
  connections in the coinbase monitoring and transaction
  validation protocols.
- Fixed transaction validation protocol to also validate
  broadcast transactions.
- Fixed broadcast monitoring protocol to stop when
  transaction are not valid anymore.

stringhandler commented:
My concern here is that the ChainMetadata is sent to all peers to advertise height. Adding in the reorg data for one use case (a connected wallet needs to know about it) means that there is a lot more data travelling the network for non-related use cases. 

Let me think on it a bit, but ideally you would only want this data to be on the network for the network calls that need it

hansieodendaal commented:
> Let me think on it a bit, but ideally you would only want this data to be on the network for the network calls that need it

Sure. There are pros and cons to both ways.

Commit:
[wallet] Update base node service to use RPC

delta1 commented:
> ideally you would only want this data to be on the network for the network calls that need it

@hansieodendaal may I suggest that this branch is rebased against #2753 - that way the `ReorgInfo` could be added to `TipInfoResponse` instead of `ChainMetadata`. I believe TipInfoResponse is only used for the console wallet and merge mining proxy.  @mikethetike thoughts?

stringhandler commented:
Agree

hansieodendaal commented:
Will do

hansieodendaal commented:
@delta1 and @mikethetike, changes applied as requested.

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -830,6 +839,39 @@ where B: BlockchainBackend
+        if let BlockAddResult::ChainReorg(_, reorg_chain) = block_add_result.clone() {
+            let tip_height = reorg_chain
+                .first()
+                .expect("Cannot extract meta data from the reorged chain")
```
Possible `panic`, better to replace with an `ok_or` or similar?

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -830,6 +839,39 @@ where B: BlockchainBackend
+            let last_reorg_best_block = Some(
+                reorg_chain
+                    .first()
+                    .expect("Cannot extract meta data from the reorged chain")
```
Same

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -830,6 +839,39 @@ where B: BlockchainBackend
+                reorg_info
+            );
+            let mut txn = DbTransaction::new();
+            txn.set_reorg_info(reorg_info);
```
This should be done inside of the same LMDB transaction as `add_block` otherwise could lead to a mismatch of data. The `mut db` prevents this currently, but the two steps are not distinct and should really be done together.

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1493,6 +1535,21 @@ fn rewind_to_height<T: BlockchainBackend>(db: &mut T, height: u64) -> Result<Vec
     db.write(txn)?;
 
+    // Update reorg info
+    let reorg_info = ReorgInfo {
```
This seems to be a duplicate of the `add_block` logic. Are you sure this is not done twice?

hansieodendaal replied:
There are different pathways/edge cases catered for here, one of which are when the entire database a base node has assembled are reorged away.

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1493,6 +1535,21 @@ fn rewind_to_height<T: BlockchainBackend>(db: &mut T, height: u64) -> Result<Vec
+    );
+    let mut txn = DbTransaction::new();
+    txn.set_reorg_info(reorg_info);
+    db.write(txn)?;
```
Should be done in the previous transaction

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1726,6 +1805,24 @@ fn restore_reorged_chain<T: BlockchainBackend>(
     db.write(txn)?;
+
+    // Update reorg info
+    if last_reorg_best_block.is_some() {
```
This is already done in `rewind_to_height`, is this duplicate code?

hansieodendaal replied:
Not duplicate code, different pathways/edge cases catered for here.

stringhandler commented on base_layer/common_types/src/chain_metadata.rs:
```diff
@@ -142,6 +142,47 @@ impl Display for ChainMetadata {
+}
+
+impl ReorgInfo {
+    pub fn new() -> Self {
```
I'd prefer to not have a Default for this struct. The struct should only exist if there was a reorg. Callers can use an `Option<ReorgInfo>` and use `None` for no reorg info. The `tip_height` of `0` is a valid value and cannot be used for detecting `None`

stringhandler replied:
The `last_reorg_best_block` can then be changed to `HashOutput` and used without checking every time

stringhandler commented on base_layer/wallet/src/base_node_service/service.rs:
```diff
@@ -223,137 +213,116 @@ where
                     );
-                },
-            }
+                    if let Err(e) = transaction_service
```
Should these not be propagated with `?`

hansieodendaal replied:
My opinion is that it is good enough to read in the log that there was an error here, as re-validation is important but not critical, and not to stop the rest of the code from executing, which is much more important/critical.

stringhandler commented on base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:
```diff
@@ -389,6 +389,7 @@ where TBackend: TransactionBackend + 'static
                             "Coinbase transaction (TxId: {}) not mined yet, still waiting.", self.tx_id,
                         );
                         // Wait out the remainder of the delay before proceeding with next loop
+                        drop(client);
```
What is the reason for dropping here?

hansieodendaal replied:
Will add a comment; this is fixing a bug where the RPC connections are not dropped but stay open.

stringhandler commented on base_layer/wallet/src/transaction_service/protocols/transaction_validation_protocol.rs:
```diff
@@ -328,6 +328,7 @@ where TBackend: TransactionBackend + 'static
                                             );
                                             e
                                         });
+                                    drop(client);
```
What is the reason for dropping here

hansieodendaal replied:
As per above, will add a comment.

stringhandler commented on base_layer/wallet/src/transaction_service/protocols/transaction_validation_protocol.rs:
```diff
@@ -348,6 +349,7 @@ where TBackend: TransactionBackend + 'static
                                     );
                                     e
                                 });
+                                drop(client);
```
What is the reason for dropping here

hansieodendaal replied:
As per above, will add a comment.

stringhandler commented on integration_tests/helpers/util.js:
```diff
@@ -153,7 +153,7 @@ function consoleLogBalance(balance) {
 }
 
-function consoleLogCoinbaseDetails(txnDetails) {
+function consoleLogTransactionDetailsFormat2(txnDetails) {
```
Let's just use one formatter. I as a developer would not know in which case to use one or the other

hansieodendaal replied:
Yip

stringhandler commented on integration_tests/features/support/steps.js:
```diff
@@ -965,24 +1008,103 @@ Then(/the number of coinbase transactions for wallet (.*) and wallet (.*) are (.
+    expect(validA).to.equal(!validB)
+});
+
+Then(/all normal transactions for wallet (.*) and wallet (.*) have consistent but opposing validity/, {timeout: 20*1000}, async function (
```
Can't these two methods be combined with a regular expression rather? There is a lot of code being duplicated here

hansieodendaal replied:
Yip, will combine

stringhandler commented on integration_tests/features/support/steps.js:
```diff
@@ -965,24 +1008,103 @@ Then(/the number of coinbase transactions for wallet (.*) and wallet (.*) are (.
+    }
+});
+
+Then(/all normal transactions for wallet (.*) are valid/, {timeout: 20*1000}, async function (
```
Same for this duplicate code

hansieodendaal replied:
Ditto

stringhandler commented on integration_tests/features/support/steps.js:
```diff
@@ -287,6 +287,21 @@ When(/I mine (\d+) blocks on (.*)/, {timeout: 600*1000}, async function (numBloc
     this.tipHeight += parseInt(numBlocks);
 });
 
+When(/I sync mine (\d+) blocks on (.*)/, {timeout: 1200*1000}, async function (numBlocks, name) {
```
What does `sync mine` mean?

hansieodendaal replied:
sync mine = mine some blocks while waiting for all other nodes to keep in sync

The problem here is these two regular expressions look similar to cucumber:
`When(/I mine (\d+) blocks on (.*) /`
       vs.
`When(/I mine (\d+) blocks on (.*) while waiting for all other nodes to keep in sync/`

Maybe this will work:
`When(/I keep all nodes in sync and mine (\d+) blocks on (.*)/`

stringhandler reviewed (commented):
I appreciate there is a lot of work in this PR, but there are a lot of cases where you are preventing the rust compiler from helping here. Any time you implement something out of the ordinary like calling `drop` or `expect`, or not propagating an error you should have a comment justifying the reason that you done so. When you do something like that, a code reviewer will immediately have questions, so you want to answer them in the comment. Obviously it is better to not use them at all, but in some cases I understand that you need to.

The approach you've used here requires the base node to keep track of the reorg and there are a few problems that can arise from this:
1. Due to a bug (current or future), this is falls out of sync and the data is junk
2. There is no incentive for base nodes to keep this data. They could also lie
3. We need to cover all scenarios in tests to make sure that this data never gets out of sync

As an alternative, what do you think about having the wallet store the header hash of the tip that it last saw, and then sending it to the base node to find out if it's still in the best chain? The advantage of this is that there are fewer logic paths in the base node implementation and no chance of ReorgInfo getting out of sync. The methods for retrieving data and finding chain splits already exist in the base node, so require less testing

philipr-za commented:
Just some thoughts on the global comments:

> 2. There is no incentive for base nodes to keep this data. They could also lie

This is true for many elements of the base node. They could lie in the chainsplit request used in the alternative below. The incentive, I think, is this approach is less work for the BN
 
> As an alternative, what do you think about having the wallet store the header hash of the tip that it last saw, and then sending it to the base node to find out if it's still in the best chain? The advantage of this is that there are fewer logic paths in the base node implementation and no chance of ReorgInfo getting out of sync. The methods for retrieving data and finding chain splits already exist in the base node, so require less testing

This approach can work but its significantly more work for the BN to do this. Instead of just recording when reorgs happen , which is a very low overhead as its done once for every reorg event, every time an individual wallet makes a request to this base node is has to retrieve a fair amount of data and do the processing to find a chain split. As opposed to just grabbing the reorg info from the db and serving it.

Commit:
Review comments:
- Simplified/consolidated cucumber tests
- Reorganozid db transactions
- Added comments to critical but not so obvious drop calls

Commit:
Review comments:

- Refactored `ReorgInfo`
